File: composeApp/src/commonMain/kotlin/dev/mks/duskread/ui/common/WaveformMeter.kt
package dev.mks.duskread.ui.common

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * A row of bars standing in for a progress track, used for both audio playback and the
 * focus timer.
 */
@Composable
fun WaveformMeter(
    progress: Float,
    modifier: Modifier = Modifier,
    barCount: Int = 18,
    seed: Int = 0,
    // "No signal", drawn as the same meter flattened to a line of dots rather than as a
    // separate empty-state ornament — see [EmptyState].
    flat: Boolean = false,
    filledColor: Color = MaterialTheme.colorScheme.primary,
    // The mockup's off-state bar (`--a-wave-off`) sits between its hairline and its
    // meta-text grey.
    dimColor: Color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.55f),
    barWidth: Dp = 2.dp,
    barGap: Dp = 1.5.dp,
) {
    val clamped = progress.coerceIn(0f, 1f)
    val width = barWidth * barCount + barGap * (barCount - 1)

    Canvas(modifier.width(width)) {
        val bar = barWidth.toPx()
        val pitch = bar + barGap.toPx()
        val radius = CornerRadius(1.dp.toPx())
        // Whole bars only.
        val filledUpTo = clamped * barCount
        val heights = waveformHeights(barCount, seed)

        for (i in 0 until barCount) {
            val barHeight = if (flat) bar.coerceAtMost(size.height) else size.height * heights[i]
            drawRoundRect(
                color = if (i < filledUpTo) filledColor else dimColor,
                topLeft = Offset(pitch * i, (size.height - barHeight) / 2f),
                size = Size(bar, barHeight),
                cornerRadius = radius,
            )
        }
    }
}

/**
 * The mockup writes each row's bar heights out by hand, and no two rows share a sequence
 * — that is the whole reason the meter reads as a clip rather than as a texture.
 */
private fun waveformHeights(count: Int, seed: Int): FloatArray {
    val steps = IntArray(count)
    var previous = -9
    for (i in 0 until count) {
        var h = seed * 0x9E3779B9.toInt() + i * 0x85EBCA6B.toInt()
        h = h xor (h ushr 15)
        h *= 0x2545F491
        h = h xor (h ushr 13)
        var step = (h ushr 1) % 11
        if (kotlin.math.abs(step - previous) < 3) step = (step + 5) % 11
        steps[i] = step
        previous = step
    }
    return FloatArray(count) { (5 + steps[it]) / 15f }
}

File: iosApp/iosApp/Components/WaveformMeter.swift
import ComposeApp
import SwiftUI

/// The per-row meter, and the "no signal" ornament when flat. Intrinsic width, not
/// stretched: `barWidth * count + gap * (count - 1)`.
struct WaveformMeter: View {
    var progress: Double
    var barCount: Int = 18
    var seed: Int32 = 0
    var flat: Bool = false
    var height: CGFloat = 15
    var filled: Color?

    private let barWidth: CGFloat = 2
    private let barGap: CGFloat = 1.5

    @Environment(\.dusk) private var dusk

    var body: some View {
        let heights = Self.heights(count: barCount, seed: seed)
        let filledUpTo = min(max(progress, 0), 1) * Double(barCount)

        HStack(alignment: .center, spacing: barGap) {
            ForEach(0..<barCount, id: \.self) { i in
                RoundedRectangle(cornerRadius: 1)
                    // Whole bars only: a half-tinted boundary bar reads as an
                    // anti-aliasing artefact at 2pt, not as sub-bar precision.
                    .fill(Double(i) < filledUpTo ? (filled ?? dusk.primary) : dim)
                    .frame(width: barWidth, height: flat ? barWidth : height * heights[i])
            }
        }
        .frame(height: height)
    }

    /// The mockup's off-state bar sits between the hairline and the meta grey.
    private var dim: Color { dusk.onSurfaceVariant.opacity(0.55) }

    /// The same integer hash the Compose meter uses, reproduced exactly.
    static func heights(count: Int, seed: Int32) -> [CGFloat] {
        var steps = [Int32](repeating: 0, count: count)
        var previous: Int32 = -9
        for i in 0..<count {
            var h = seed &* Int32(bitPattern: 0x9E37_79B9) &+ Int32(i) &* Int32(bitPattern: 0x85EB_CA6B)
            h ^= Int32(bitPattern: UInt32(bitPattern: h) >> 15)
            h = h &* 0x2545_F491
            h ^= Int32(bitPattern: UInt32(bitPattern: h) >> 13)
            var step = Int32(bitPattern: UInt32(bitPattern: h) >> 1) % 11
            if abs(step - previous) < 3 { step = (step + 5) % 11 }
            steps[i] = step
            previous = step
        }
        return steps.map { CGFloat(5 + $0) / 15 }
    }
}
